File: 2023/python/day_03.py
# https://adventofcode.com/2023/day/3

INPUT_FILE: str = "2023/data/day_03.txt"
INPUT_FILE_SMALL: str = "2023/data/day_03_small.txt"


def parse_line(line: str) -> tuple[list[tuple[int, int, int]], list[tuple[str, int]]]:
    nums, symbols = [], []
    numstr = ""
    start, end = -1, -1
    for i, ch in enumerate(line + "."):
        if ch in "0123456789":
            end = i
            if start == -1:
                start = end
            numstr += ch
        else:
            if numstr != "":
                nums.append((int(numstr), start, end))
                numstr, start, end = "", -1, -1
            if ch == ".":
                numstr, start, end = "", -1, -1
            else:
                symbols.append((ch, i))
    return nums, symbols


def get_nums_and_symbols_coords(
    lines: list[str],
) -> tuple[list[tuple[int, int, int, int]], list[tuple[str, int, int]]]:
    # parsing twice, but the number of lines is small (~ 150)
    all_nums = [
        (value, row, start, end)
        for row, line in enumerate(lines)
        for value, start, end in parse_line(line)[0]
    ]

    all_syms = [
        (sym, row, col)
        for row, line in enumerate(lines)
        for sym, col in parse_line(line)[1]
    ]
    return all_nums, all_syms


def is_adjacent(number_info, symbol_info):
    _, number_row, number_start_col, number_end_col = number_info
    _, symbol_row, symbol_col = symbol_info

    return (
        number_row - 1 <= symbol_row <= number_row + 1
        and number_start_col - 1 <= symbol_col <= number_end_col + 1
    )


def get_input(input_file: str) -> list[str]:
    with open(input_file, "r") as f:
        return [x.strip() for x in f.readlines()]


def solve_day_03_part_1(input_file: str = INPUT_FILE):
    lines = get_input(input_file)
    nums, symbols = get_nums_and_symbols_coords(lines)
    result = 0
    for num in nums:
        for symbol in symbols:
            if is_adjacent(num, symbol):
                result += num[0]
    return result


def solve_day_03_part_2(input_file: str = INPUT_FILE):
    lines = get_input(input_file)
    numbers, symbols = get_nums_and_symbols_coords(lines)
    result = 0

    for symbol in symbols:
        if symbol[0] != "*":
            continue
        adj_list = []
        for number in numbers:
            if is_adjacent(number, symbol):
                adj_list.append(number[0])
        if len(adj_list) == 2:
            result += adj_list[0] * adj_list[1]

    return result


if __name__ == "__main__":
    print(f"Solution for Part 1 (small): {solve_day_03_part_1(INPUT_FILE_SMALL)}")
    print(f"Solution for Part 2 (small): {solve_day_03_part_2(INPUT_FILE_SMALL)}")

    print(f"Solution for Part 1: {solve_day_03_part_1(INPUT_FILE)}")
    print(f"Solution for Part 2: {solve_day_03_part_2(INPUT_FILE)}")

File: 2023/rust/src/day_03.rs
// https://adventofcode.com/2023/day/3

use std::fs::File;
use std::io::Read;
use std::path::Path;

struct Number {
    value: i32,
    start: i32,
    end: i32,
    row: i32,
}

struct Symbol {
    value: char,
    row: i32,
    col: i32,
}

fn read_file_contents(input_fname: impl AsRef<Path>) -> (Vec<Number>, Vec<Symbol>) {
    let file_path = input_fname.as_ref();
    let mut input = File::open(file_path).expect("Failed to open file");
    let mut contents = String::new();
    let _ = input
        .read_to_string(&mut contents)
        .expect("Failed to read file");

    let mut numbers: Vec<Number> = Vec::new();
    let mut symbols: Vec<Symbol> = Vec::new();

    let mut row = 0;

    for line in contents.lines() {
        row += 1;
        let mut col = 0;
        let extended_line = format!("{}{}", line, ".");
        let mut crt_number = String::new();
        let mut num_start = -1;
        let mut num_end = -1;
        for c in extended_line.chars() {
            col += 1;

            if c.is_digit(10) {
                crt_number.push(c);
                num_end = col;
                if num_start == -1 {
                    num_start = col;
                }
            } else {
                if crt_number.len() > 0 {
                    numbers.push(Number {
                        value: match crt_number.parse::<i32>() {
                            Ok(value) => value,
                            Err(_) => panic!("Invalid number {}", crt_number),
                        },
                        start: num_start,
                        end: num_end,
                        row: row,
                    });
                }

                if c != '.' {
                    symbols.push(Symbol {
                        value: c,
                        row: row,
                        col: col,
                    });
                }

                crt_number = String::new();
                num_start = -1;
                num_end = -1;
            }
        }
    }

    (numbers, symbols)
}

fn is_adjacent(number: &Number, symbol: &Symbol) -> bool {
    (number.row - symbol.row).abs() <= 1
        && number.start - 1 <= symbol.col
        && symbol.col <= number.end + 1
}

pub fn solve_day_03_part_1(input_fname: impl AsRef<Path>) -> i32 {
    let (numbers, symbols) = read_file_contents(input_fname);
    let mut result = 0;

    for number in &numbers {
        for symbol in &symbols {
            if is_adjacent(number, symbol) {
                result += number.value;
            }
        }
    }

    result
}

pub fn solve_day_03_part_2(input_fname: impl AsRef<Path>) -> i32 {
    let (numbers, symbols) = read_file_contents(input_fname);
    let mut result = 0;

    for symbol in &symbols {
        if symbol.value != '*' {
            continue;
        }
        let mut first = -1;
        let mut second = -1;
        let mut more_than_two = false;
        for number in &numbers {
            if is_adjacent(number, symbol) {
                if first == -1 {
                    first = number.value;
                } else {
                    if second == -1 {
                        second = number.value;
                    } else {
                        more_than_two = true;
                    }
                }
            }
        }
        if more_than_two {
            continue;
        }
        if first != -1 && second != -1 {
            result += first * second;
        }
    }

    result
}
